system/datatypes/mailinglist.php

```diff
@@ -442,6 +452,77 @@ private function unsubscribeAction(HyphaRequest $request) {
 		return ['redirect', isset($_GET['redirect']) ? $_GET['redirect'] : $this->path()];
 	}
 
+	/**
+	 * @param HyphaRequest $request
+	 * @return null|array
+	 */
+	private function remindAction(HyphaRequest $request) {
+		$code = isset($_GET['code']) ? $_GET['code'] : null;
+		if (null == $code) {
+			notify('error', __('missing-arguments'));
+			return null;
+		}
+
+		$this->xml->lockAndReload();
+
+		$xpath = './/' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_CONFIRM_CODE . '=' . xpath_encode($code) . ' and @' . self::FIELD_NAME_STATUS . '!="' . self::ADDRESS_STATUS_UNSUBSCRIBED . '"]';
+		$address = $this->findAddresses($xpath)->first();
+
+		if (!$address instanceof HyphaDomElement) {
+			$this->xml->unlock();
+			notify('error', __('invalid-code'));
+			return null;
+		}
+
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) !== self::ADDRESS_STATUS_PENDING) {
+			$this->xml->unlock();
+			notify('error', __('ml-no-reminder-address-no-longer-pending'));
+			return null;
+		}
+
+		if ($address->getAttribute(self::FIELD_NAME_REMINDED) != true) {
+			$email = $address->getAttribute(self::FIELD_NAME_EMAIL);
+			$address->setAttribute(self::FIELD_NAME_REMINDED, true);
+			$this->xml->saveAndUnlock();
+
+			$this->sendConfirmationMail($code, $email);
+		}
+
+		notify('success', ucfirst(__('ml-reminder-sent')));
```

Comment:
We've discussed this before, whether to show an error or success when the wanted state is already the current state, but in this case I really wonder if this is the right approach. Saying "reminder sent" explicitly indicates that something actually happened. In the case that something is wrong and e-mails are not received, it really does not help to have additional misunderstanding about whether something was even sent in the first place.

I realize that this is probably a corner case because the reminder button will not normally be shown when a reminder was already sent, but users will probably manage to keep open an old interface somewhere and then get confused about the message sent.

Reply:
This is indeed an edge case. The remind action can and should only be triggered one time. If the remind action is triggered a second time, there is no need to send another reminder, and also no need to inform the user that this edge case occurred, because for the user the end result is the same.

Reply:
I've updated the code, but still don't think it's needed.

Reply:
> because for the user the end result is the same.

I think this entirely depends on the timeframe. If two users click the remind button at the same time, then the result is pretty much the same, yes: A reminder has been sent. However, if more time passes, then the message might cause a user to (rightfully) assume that *another* reminder has been sent (and e.g. tell the reminded user that he should have received another e-mail, which is then untrue). And even when both presses occur in short succession, this might lead user to believe that *two* reminder emails have been sent, which is not actually the case.